Shared stream

tmccombs opened the pull request:
This is similar to #1217 but uses an underlying boxed slice to buffer items, and as a side-effect has a finite buffer size. 

This is a work in progress. I'd still like to add some more rigorous unit tests, but I'd like feedback on my general approach, and have other eyes check my logic.

tmccombs committed:
Refactor Notifier out of future/shared

So that we can also use it for stream/shared

tmccombs committed:
Implement a shared Stream

ebkalderon commented:
Has there been any additional progress on this recently? I happen to be looking for this exact feature.

tmccombs commented:
I haven't done anything more with this. I would like some feedback on my approach before I put more effort into polishing it.

tmccombs commented:
@cramertj, @MajorBreakfast any thoughts on this?

ebkalderon commented:
Is this PR still being worked on? I ran into several more cases over the past two years where this API could be useful. I'm willing to take over this PR and help drive it to completion if the original author is away. @cramertj, any thoughts on this?

tmccombs commented:
I'm still around. It's been a while since I looked at this, but I was waiting on feedback from the maintainers.

tmccombs commented:
Looks like the shared `future::Shared` was refactored a little bit. So I rebased my branch on master. One piece that is still missing from this is some unit tests for `stream::Shared`. If you wanted to help write some @ebkalderon that might be helpful.

ebkalderon commented:
Sure thing, @tmccombs! I'll write a few, perhaps based on the tests for `future::Shared`, and I'll post them here.

ebkalderon commented:
@tmccombs I didn't get around to writing tests for this, unfortunately, but it seems that the current commit is failing in CI due to a `clippy` lint for `unnested-or-patterns` ([source](https://travis-ci.com/github/rust-lang/futures-rs/jobs/350251022#L340)). I can write some tests once this is fixed and open a PR against your branch, perhaps, so you can merge them in and have them ready for review here.

tmccombs commented:
@ebkalderon  those clippy lints are for code I didn't touch, and making the suggested change results in an "or-patterns syntax is experimental" error on stable rust.

ebkalderon commented:
Got it! Thanks for clarifying and updating the branch. I'll get started writing some tests and open a PR against your fork.

tmccombs commented:
Conflicts have been resolved and this is up-to-date with current master.

tmccombs commented:
The failed miri test looks like it is probably unrelated, but I'm not completely sure.

XLordalX commented:
Any updates?

tmccombs commented:
@XLordalX your comment got me to revisit this, and I wrote some tests, which led to finding some bugs which I have now fixed, and reworked it a fair amount. 

It is still failing CI. Something about a race condition found with thread sanitization. I don't really understand it yet. 


I would kind of like some feedback from the futures maintainers to know if this a direction worth pursuing. And, to be honest, I kind of hope someone comes along and says there is a simpler way to do it.

tmccombs committed:
Add more tests for shared streams

And fix bug with wrong operator

tmccombs committed:
Fix race condition in shared stream

tmccombs committed:
Fix data races in shared stream

I don't completely understand why the stronger ordering is needed, but
it seems to fix the data races detected by the thread sanitizer.

tmccombs committed:
Add more tests for shared streams

tmccombs committed:
Don't increment slot for terminated shared stream

XLordalX commented:
@taiki-e You able to look at this?

tmccombs committed:
Fix ordering for loading state of next slot while filling in shared stream

tmccombs commented:
hmm, looks like builds are currently broken on master.

tmccombs committed:
Fix builds for shared streams

tmccombs committed:
Fix conditions for wakerset
